docs/d-attributes/fattr.md

```diff
@@ -0,0 +1,564 @@
+
+Now let us annotate the main function definition with `@safe`:
+```d
+void main() @safe
+```
+The compiler correctly highlights:
+```d
+test.d(10): Error: cannot take the address of local var in @safe function main
+```
+
+*Practice*: Navigate to the `demo/safe` directory. Inspect the source file. Compile and run the code.
+
+1. What does the code do? Why is it useful to take the address of a parameter?
+2. Add the `@safe` attribute to the `main` function. What happens?
+3. Add `@safe` to the `func` and `gun` functions. Analyze the error messages.
+4. How can we get rid of the first error message?
+5. What about the second error message?
+
+*Practice*: Navigate to the `demo/code-safety` directory. Inspect the source file. Compile and run the code.
+
+1. Add `@safe` to the main function and then make the code compile by solving the issues.
+
+### `@trusted`
+
+A `@trusted` function provides the same guarantees as a `@safe` function, but it is assumed that the checks were done manually by the programmer.
+That is because sometimes, even though the `@safe`ty rules work well to prevent memory corruption, they prevent a lot of valid, and actually safe code.
+For example, consider a function that wants to use the system call `read`, which is prototyped like this:
+```d
+ssize_t read(int fd, void* ptr, size_t nBytes);
+```
+
+For those unfamiliar with this function, it reads data from the given file descriptor, and puts it into the buffer pointed at by `ptr` that is expected to be `nBytes` bytes long.
+The function returns the number of bytes actually read, or a negative value if an error occurs.
+However, this C idiom of passing the buffer and length separately is not a safe pattern because it's easy to pass a length that exceeds the buffer llength.
+```c
+char *buf = malloc(128);
+auto nread = read(fd, buf.ptr, 10000);
+```
+This is clearly an unsafe pattern, so in D it is not possible to mark this code as `@safe`.
+
+To solve this situation, D provides the `@trusted` attribute, which tells the compiler that the code inside the function is assumed to be `@safe`, but will not be mechanically checked.
+It's on you, the developer, to make sure the code is actually `@safe`.
+
+A function that solves the problem might look like this in D:
+```d
+auto safeRead(int fd, ubyte[] buf) @trusted
+{
+    return read(fd, buf.ptr, buf.length);
+}
+```
+
+Before marking an entire function with `@trusted`, consider first if calling it from a given context would compromise the memory safety of the program.
+If this is the case, this function should not be marked `@trusted` under any circumstances.
+Even if the intention is to only call it in safe ways, the compiler will not prevent unsafe usage by others.
+In our case, `safeRead` should be fine to call from any `@safe` context, so it's a great candidate for the `@trusted` attribute.
+
+*Practice*: Navigate to the `demo/trusted` directory. Inspect the source file. Compile and run the code.
```

Comment:
```suggestion
### Practice

1. Navigate to the `demo/trusted` directory. Inspect the source file. Compile and run the code.
```

Reply:
Now, the bullets from 1-3 must be a part of the this ex 1. here.

The next `*Practice*:` disappears, the "Navigate to bla bla" becomes ex 2. and the next 1-3 points become a part of ex 2